ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -34,6 +36,91 @@ module Winbind = struct
     debug "init_service To be implemented in CP-36085"
 end
 
+let net_cmd = "/usr/bin/net"
+
+let from_config ~name ~err_msg ~config_params =
+  match List.assoc_opt name config_params with
+  | Some v ->
+      v
+  | _ ->
+      raise (Auth_service_error (E_GENERIC, err_msg))
+
+let assert_hostname_valid () =
+  let hostname =
+    (fun __context ->
+      Helpers.get_localhost ~__context |> fun host ->
+      Db.Host.get_hostname ~__context ~self:host)
+    |> Server_helpers.exec_with_new_task "retrieving hostname"
+  in
+  let all_numbers hostname =
+    String.fold_left (fun b ch -> b && ch >= '0' && ch <= '9') true hostname
+  in
+  match all_numbers hostname with
+  | true ->
+      raise
+        (Auth_service_error
+           ( E_GENERIC
+           , Printf.sprintf "hostname '%s' cannot contain only digits." hostname
+           ))
+  | false ->
+      ()
+
+let assert_domain_equal_service_name ~service_name ~config_params =
+  (* For legeacy support, if domain exist in config_params, it must be equal to service_name *)
+  let domain_key = "domain" in
+  if List.mem_assoc domain_key config_params then
+    let domain = List.assoc domain_key config_params in
+    if domain <> service_name then
+      raise
+        (Auth_service_error
+           (E_GENERIC, "if present, config:domain must match service-name."))
+
+let extract_ou_config ~config_params =
+  try
+    let ou = from_config ~name:"ou" ~err_msg:"" ~config_params in
+    ([("ou", ou)], [Printf.sprintf "createcomputer=%s" ou])
+  with Auth_service_error _ -> ([], [])
+
+let query_domain_workgroup domain =
+  let key = "Pre-Win2k Domain" in
+  let lines =
+    Helpers.call_script ~log_output:On_failure net_cmd
+      ["ads"; "lookup"; "-S"; domain]
+  in
+  match Xapi_cmd_result.of_output_opt ~sep:':' ~key ~lines with
+  | Some v ->
+      v
+  | None ->
+      raise (Auth_service_error (E_LOOKUP, "Failed to lookup domain workgroup"))
+
+let config_winbind_damon domain workgroup =
+  let open Xapi_stdext_unix in
+  let smb_config = "/etc/samba/smb.conf" in
+  let conf_contents =
+    String.concat "\n"
+      [
+        "# autogenerated by xapi"
+      ; "[global]"
+      ; "kerberos method = secrets and keytab"
+      ; Printf.sprintf "realm = %s" domain
+      ; "security = ADS"
+      ; "template shell = /bin/bash"
+      ; "winbind offline logon = Yes"
+      ; "winbind refresh tickets = Yes"
+      ; "winbind enum groups = no"
+      ; "winbind enum users = no"
```

Comment:
To handle complex inter-domain/forest trusts (like our own CITRITE) I think you need this setting too:
```
winbind use krb5 enterprise principals = yes
```

This should allow you to join the machine to FOO.EXAMPLE.COM, but login with EXAMPLE.COM credentials if trust is properly set up between FOO.EXAMPLE.COM and EXAMPLE.COM (e.g. because you only have logins on EXAMPLE.COM, and nobody has logins on FOO.EXAMPLE.COM, but EXAMPLE.COM doesn't allow joining machines, but FOO.EXAMPLE.COM does)

Reply:
The document https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html says:
```
It is not really required to know about the trust topology. We can just rely on the [K]DCs of our primary domain (e.g. PRIMARY.A.EXAMPLE.COM) and use enterprise principals e.g. upnfromB@B.EXAMPLE.COM@PRIMARY.A.EXAMPLE.COM and follow the WRONG_REALM referrals in order to find the correct DC. The final principal might be userfromB@INTERNALB.EXAMPLE.PRIVATE.
```
But set this to `yes` looks like make things better ?

Reply:
Yes I think trust topology is complicated and `winbindd` may not be able to get the full picture, so when you set that config option it'll rely on the (K)DC servers telling it where it should go to authenticate the user.
I was not able to join a machine and authenticate to our internal AD without this setting. Have you tried with and without this setting?

Reply:
I will follow the suggestion and give it a try

Reply:
I had a tried and updated the details in the channel.